text/0208-ephemeral-resources.md

```diff
@@ -0,0 +1,78 @@
+
+
+## Internal details
+
+### ResourceProvider
+
+#### NewResourceProvider([resource], [validator]) ResourceProvider
+
+NewResourceProvider instantiates an implimentation of the ResourceProvider interface. As argumentes, it optionally takes an initial set of resource attributes, and a validator. 
+
+The ResourceProvider interface has the following methods
+
+#### MergeResource(resource)
+
+MergeResource creates a new resource, representing the union of the resource parameter and the resource contained within the Provider. The ResourceProvider holds a reference to the nwe resource.
+
+#### SetAttribute(key, value)
+
+SetAttribute functions the same as MergeResource, but only adds a single attribute.
+
+#### GetResource() Resource
+
+GetResource returns a reference to the current resource held by the ResourceProvider.
+
+#### FreezePermanent()
+
+FreezePermanent is called by the SDK one it has been stared. After FreezePermanent has been called, any calls to MergeResource or SetAttributes will only be applied if the validator acceptes the input.
+
+#### Implementation Notes
+
+For multithreaded systems, a lock should be used to queue all calls to MergeResource and SetAttribute. But the resource reference held by the ResourceProvider should be updated atomically, to prevent calls to GetResource from being blocked.
+
+### SDK Changes
+
+NewTraceProvider, NewMetricsProvider, and NewLogProvider now take a ResourceProvider as a parameter. For backwards compatibility, the Resource parameter remains functional. If both a resource and a resource provider as passed as parameters, the resource is merged into the ResourceProvider, then discarded.
+
+FreezePermanent is then called by the provider.
+
+Internally, providers hold a reference to the ResourceProvider, rather than a specific resource. When creating a signal, such as a span, metric, or log, GetResource() is called to obtain a reference to the correct resource to attach to the signal.
+
+## Trade-offs and mitigations
+
+This change should be fully backwards compatible, with one potential exception: fingerprinting. It is possible that an analysis tool which accepts OTLP may identify individual services by creating an identifier by hashing all of the resource attributes. 
+
+In this case, it is recommended that these systems modify their behavior, and choose a subset of permanent resources to use as a hash identifier.
+
+## Prior art and alternatives
+
+An alternative to ephemeral resources would be to create span, metrics, and log processors which attach these ephemeral attributes to every instance of every signal. This would not require a modification to the specification.
+
+There are two problems to this approach. One is that the duplication of attributes is very inefficient. This is a problem on clients, which have a limited newtwork bandwidth. This problem is compounded by a lack of support for gzip and other compression algorithms on the browser.
```

Comment:
It would be great to quantify this. How inefficient is it? A benchmark demonstrating this would be a strong argument in favour of the proposed approach.

Reply:
If processors can change scope attributes, they might be a good candidate to solve this as well. 

> lack of support for gzip and other compression algorithms on the browser

I'm not an expert on browser stuff but can you expand on this? On its surface it seems wrong since gzipped static resources show up everywhere on the internet and there are js implementations of gzip (like [this](https://github.com/nodeca/pako)). This [stackoverflow](https://stackoverflow.com/questions/424917/why-cant-browser-send-gzip-request) post suggests that a part of it is because a browser client can't know if the server can accept gzipped data, but [OTLP requires gzip support](https://github.com/open-telemetry/opentelemetry-specification/blame/main/specification/protocol/otlp.md#L78).

Reply:
Never mind on the OTLP gzip support, I see it says that clients MAY gzip the content.

Reply:
@tigrannajaryan Regarding the limited network bandwidth, the sendBeacon() API has a payload limit of 64KB. Assuming session.id attribute that looks like this when sent over the wire

```
{"key":"session.id","value":{"stringValue":"8fded6726f630a327ee3be41174a8a91"}}
```

It adds 79 bytes per each signal. The number of spans/events per export will depend on the type of application and which instrumentations are present. But assuming that 100 is plausible, this adds almost 8kB to the payload.

This will further increase if we add additional context attributes (user attributes, URL etc.).

Reply:
> I think there is also a danger of an attack on the server - compressed data could be expanded to a very large content. 

I'm not familiar with this type of attack, but surely a server decompressing a payload can have logic that cuts it off after it exceeds some threshold? 

> And lastly gzip compression is not native to browsers, so there is CPU overhead, which is important to consider for impact on user experience, especially when sending data while the page is unloading.

There's always going to be CPU overhead with compression, but as a rule of thumb applications tend to be bounded by network rather than CPU, in which case the user experience is improved because it takes less time to compress and send a smaller payload than it does to send an uncompressed larger payload. Its for this reason that its standard to gzip static resources returned to browser applications - the page loads faster despite spending CPU decompressing the payload because the app network is the bound. 

I wrote a little test suite to explore some scenarios related to compression and serializing with OTLP json vs the binary format. The client instrumentation group is obviously interested in minimizing payload sizes, primarily through reducing the amount of duplicated data that is sent. I'm not necessarily opposed to this, but the real gains in reducing payload size are going to come if the group embraces compression and does the work needed to serialize using the binary format instead of json. The test suite I wrote shows this:

| Scenario | JSON Size Raw | JSON Size Compressed (% reduction) | Binary Size Raw (% reduction) | Binary Size Compressed (% reduction) |
|----------|---------------|----------------------|-----------------|------------------------|
| session id on resource - 1 record | 531 | 300 (177.00) | 162 (327.78) | 153 (347.06) |
| session id on resource - 10 records | 2934 | 325 (902.77) | 703 (417.35) | 160 (1833.75) |
| session id on resource - 100 records | 26964 | 451 (5978.71) | 6103 (441.82) | 201 (13414.93) |
| session id on span - 1 record | 530 | 305 (173.77) | 164 (323.17) | 153 (346.41) |
| session id on span - 10 records | 3977 | 335 (1187.16) | 1290 (308.29) | 166 (2395.78) |
| session id on span - 100 records | 38447 | 520 (7393.65) | 12540 (306.59) | 247 (15565.59) |

Explanation:
* Compared including session id on resource vs on span records for payloads of 1, 10, and 100 spans. Source code lives [here](https://github.com/jack-berg/opentelemetry-java/blob/gzip-test/exporters/otlp/all/src/test/java/io/opentelemetry/exporter/otlp/GzipTest.java). 
* The "JSON Size Raw" column shows the payload size of json encoded OTLP. Comparing the session id on resource rows to the session id on span rows, there's gains to be had by including the session id on the resource, especially as the number records grows.
* The "JSON Size Compressed" column shows the size of the payloads when the JSON is compressed. Notice the dramatic % reduction versus raw json, and how the difference between including session id on resource and session id on span is much more modest. For example with 10 records including the session id on spans results in only a 10 byte increase in payload size.
* The "Binary Size Raw" and "Binary Size Compressed" show the same size of raw and compressed binary encoded OTLP. Notice that the Binary Size Raw for "session id on span - 100 records" is more than 2x smaller (12540) than the JSON Size Raw for "session id on resource - 100 records" (26964). The interpretation is that setting aside compression, pursuing binary encoding yields bigger payoffs than moving session id to the resource and keeping json. Compression also yields good results on binary payloads, and the best solution would be binary + gzip. 

Again, I'm not necessarily opposed to this idea of ephemeral resource attributes (need to do some thinking of whether anything else depends on the immutability of resources before I form an opinion). Just want to make the point that compression and binary encoding are going to move the needle in a much more serious way.

Reply:
Thanks for tests @jack-berg 

To confirm, for 100 spans, putting the session id in the resource results in 451 bytes of compressed JSON payload, putting the session id in the span results in 520 bytes of compressed JSON payload. Do I read the results correctly?

Reply:
@jack-berg @tigrannajaryan there's no denying that compression would beat all the contestants by a great margin. To design OTel for the long term, it would be good to design for the superset of use-cases. We want to design even for the environments that cannot afford the additional download size for the compression library and the additional CPU usage for performing the compression. 

Using scope attributes for the purpose was a good idea - iff only it was possible to manage them centrally, like a ResourceProvider in this proposal. Currently, I believe, they are managed by the individual instrumentation libraries separately. Scope attributes has an interesting property that they are equivalent to signal attributes - this allows for one to start sending the attribute on each signal instance and optimize to send once at the scope level if required.  Similarly, we could send the ephemeral attributes at the signal level now iff you can show a path to send them at the top level later.  Resource attributes do not have that property - we are seeking for it with the ephemeral attributes of Resource.

Reply:
Thank you @t2t2 for the detailed explanation! And thank you @jack-berg and @tigrannajaryan for the research.

I do not pretend to be a browser expert, but @t2t2 explains the situation as I have heard it from others.

Basically, you can load assets into the browser using compression because the browser unpacks these for you in a multithreaded C++ environment. But when sending requests using XHR, there is no way to hand off your payload to the underlying runtime where it can be compressed using the browser's gzip implementation. So you have to bring your own gzip implementation, and then block the UI thread whenever you compress. 

As a result, no current RUM clients today use compression. Instead, they use short codes and other tricks to minimize data size, and design their protocols to avoid any unnecessary  duplication.

Longer term, it looks like we would benefit from building a new browser SDK that is separated further from the current node SDK, to minimize load times and other issues. And it may be beneficial to design a new protocol specifically tailored for use in the browser - sending data over the open internet on a low bandwidth cellular connection has enough issues that this may be beneficial for reasons beyond just compression.

All that aside: back to ephemeral resources. I present the efficiency issue here because it was raised in the Client SIG as a blocker to storing ephemeral resources as span and event attributes, even as a temporary solution.

But! While I think all of this is worth mentioning, I would prefer that the discussion of ephemeral resources not get too hung up on efficiency. I believe there is benefit in designing our protocol to reflect our intended data model. 

We could just shove these resources into other scopes which happen to be available. Heck, we could just do away with resources entirely, shove all of the resources into span attributes, and let compression take care of it... but we would lose an important chunk of out data model in the process, and that makes consuming OTLP very difficult. We benefit from creating different envelopes like resources, instrumentation scopes, and spans, and only putting attributes on those envelopes which match the scope of that specific envelope. 

In fact, even more scopes would be better. For example, we currently lack the concept of trace-level attributes; as a result, our end users abuse baggage in odd ways to fake trace-level attributes by shoving the attributes into baggage and then... you guessed it... slapping them as span attributes onto every span. And, even though it "works," they aren't too happy about it, because it feels clunky and hackish. So I think we should be open to the idea that we haven't yet finished with the design of OTLP.

And, for that reason, I would ask that we judge ephemeral attributes by this measure: how do we correctly and intentionally model attributes which apply process-wide, to all telemetry emitted by an SDK? Intriguingly, precisely this scope is already present in OTLP - resources. We just have a restriction on resources which turns out to be untenable, which we discovered when we expanded OTel from just representing servers to also representing clients.

So, that it my case for this proposal. But! I am open to other approaches! For example, if we feel that this extension breaks a compatibility guarantee (and I don't believe it does, but I could be proven wrong), then we could add ephemeral resources to OTLP as it's own separate attribute set. I didn't propose that option because I felt it was messier than extending our current definition of resources to account for resources which represent attributes that change over time. 

In any case, I ask the spec community to consider this: how do we correctly model the scope of these attributes, rather than just shove them into other scopes which happen to be lying around? These attributes appear to be a fundamental enough part of modeling clients that they should be given a proper place within our data model.

Thanks again for the discussion, and the attention to this issue. :)

Reply:
>JSON's data size could also gain advantages by creating an alternative of otlp/json with one-two letter object keys which could turn compression wins < extra impact of having to load a compression library.

Exactly.

The current OTLP/JSON is extremely inefficient without compression. If we believe uncompressed JSON is going to be used in production we need a much better JSON structure. It can be done via shortened keys (as mentioned above), by using dictionary encoding (eliminates duplicate strings), etc. Such targeted and comprehensive format optimization will likely have far more effect on the payload size than the single deduplication change proposed by this OTEP.

Let me reiterate: if the goal of this OTEP is to have more efficient uncompressed JSON representation then I think we can do better without the need to introduce new ephemeral attributes concept. Instead a more efficient and compact OTLP/JSON format is possible. If that's the goal now is a great time to do it since OTLP/JSON is on the verge of being declared stable (but it is not yet, so we are free to do it).

Reply:
> if the goal of this OTEP is to have more efficient uncompressed JSON representation

I consider session and page url as the context for events/spans emitted and I thought it is a reasonable ask to have a normalized representation and not repeat them. It will really be a good design that does not rely on compression.

I believe the hesitation for this otep is due to the complexity of the changes proposed and the unknown implications. I propose adding an entirely new field called `ephemeral_resource` as a sibling to `resource` in `ScopeSpans` and `ScopeLogs` - this way, the original `resource` remains immutable and the new field can be use for the ephemeral attributes of the resource. @tedsuo

Reply:
> While decompressing is common in browsers (be it gzip or brotli), none of the current request APIs expose a way to have browser compress the request (MDN: [XHR](https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest), [fetch](https://developer.mozilla.org/en-US/docs/Web/API/fetch))

This is incorrect.  The [CompressionStreams API](https://wicg.github.io/compression/) provides a native solution for this and is [supported in Chromium-based browsers](https://developer.mozilla.org/en-US/docs/Web/API/Compression_Streams_API#browser_compatibility) already.

> This does mean that indeed you have to bring your own compression methods. More code = larger bundle that the browser needs to download, parse and execute. First phase is network bound (but does benefit from compression itself), while second and third are CPU bound. 

Also benefits from caching.  In a network-constrained situation the cost of retrieving the additional code is paid once and the result cached.  Conditional requests and etags are your friend.

> Also in most cases instrumentation is required to be loaded ASAP (sometimes even before rest of the content on the page), causing site loading to be blocked until code is downloaded (should it be in the <head> of the page)

The additional code for compression is only needed to export telemetry and does not need to be loaded at the same time as the code enabling instrumentation.  Deferring until an export is required can increase the time-to-export but would not impact time-to-interaction or any other user-focused timing.

>  I propose adding an entirely new field called ephemeral_resource as a sibling to resource in ScopeSpans and ScopeLogs - this way, the original resource remains immutable and the new field can be use for the ephemeral attributes of the resource.

This is inverted.  `ResourceSpans` contain `ScopeSpans`, not the other way around.  Ephemeral resource attributes could be added as `Scope*` attributes on each `Scope*` produced during the time when the ephemeral resource attributes are active, but I'm not sure I see how changing the OTLP data structure advances the conversation in a safe way.  That is the most invasive way of going about this that I could think of.

Reply:
@Aneurysm9 sorry my bad, I meant `ResourceSpans` and `ResourceLogs` and not `ScopeSpans` and `ScopeLogs`. I corrected this in my previous comment, can you check if it makes sense this time?

Reply:
The cost to generate, serialize, and compress that many spans is also not a synchronous process that takes x milliseconds, but many small processes which each take a small fraction of X. It is most important to ensure that each individual step doesn't impact user experience. With the example of 100 spans otlp -> protobuf -> pako on the pixel 4a given, the whole process is 4.393ms but you have 2 chances to yield to the event loop to ensure user experience is not affected.

Reply:
> This is incorrect. The [CompressionStreams API](https://wicg.github.io/compression/) provides a native solution for this and is [supported in Chromium-based browsers](https://developer.mozilla.org/en-US/docs/Web/API/Compression_Streams_API#browser_compatibility) already.

Have missed it but I generally don't consider new browser features as a solution unless usage% is >90% ([and well, safari has a monopoly on ios so....](https://caniuse.com/?search=CompressionStream)) (also 90% is probably low considering how much RUM products are asked for IE11 support but they already have a miserable experience due to using IE in current year so making it optional is worth consideration)

> but you have 2 chances to yield to the event loop to ensure user experience is not affected.

There is one but - not when user is leaving the page, tho generally you don't have 100 spans then, making it a question of how much do you want to maintain 2 different code paths (a sync one and an "async" one)

Reply:
@tigrannajaryan i just want to emphasize what @scheler said, that the purpose of this OTEP is not to avoid compression or gain efficiency, but to extend our data model in a way that correctly represents these attributes. 

If we don't want to extend the current Resource concept, we could add a new concept, call it ProccessScope or something similar, and have it work in effectively the same manner. 

Personally, I'd prefer we extend resources over adding a new scope. But I prefer both over an approach that makes it impossible to cleanly implement RUM using OpenTelemetry.

In other words, I'm against "just tack on the process scope as span/event attributes" the same way I'd be opposed to "just tack on the instrumentation scope as span/event attributes." In both cases, yes it would "work." But it would create a headache for implementers and confusion for users. 

We should strive for a clean data model, where everything is explained just by looking at the data structure.